### PR TITLE
transform: luaxform transform script

### PR DESCRIPTION
**Pull request by jlucovsky**
Continuation of #11876  

Add the `luaxform` transform that uses a Lua script to receive and transform a sticky buffer.

Lua transforms receive arguments
- (always): length of input buffer and input buffer
- (always): number of additional arguments (may be 0)
- (when present): argument array (bounded by argument count)

Arguments are passed to the lua script as they appear in the rule:
```
 luaxform:./lua/lua-transform.lua, bytes 4, offset 13, hash sha256;
 ```
 When the transform is invoked, 3 additional arguments are passed (arg count will be 3)
- `bytes 4`
- `offset 13`
- `hash sha256`

The Lua script is responsible for parsing and using the arguments and returning a tuple: `output-buffer, output-buffer-byte-count`

I used this lua script:
```
function init (args)
    local needs = {}
    return needs
end

function transform(input_len, input, argc, args)
    SCLogNotice("Input length: " .. input_len .. " arg-count: " .. argc)
    for i = 1, argc do
        print("Argument " .. i .. ":", args[i])
    end
    return string.upper(input), input_len
end

return 0
```

Example rule using the lua script:
```
alert http any any -> any any (msg:"Lua Xform example"; flow:established;  file.data; luaxform:./lua/lua-transform.lua; content: "abc"; sid: 1;)
```
Link to ticket: https://redmine.openinfosecfoundation.org/issues/2290

Describe changes:
- Luaxform transform changes

Updates
- Documentation updates per review

### Provide values to any of the below to override the defaults.

- To use an LibHTP, Suricata-Verify or Suricata-Update pull request,
  link to the pull request in the respective `_BRANCH` variable.
- Leave unused overrides blank or remove.

SV_REPO=
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2044
SU_REPO=
SU_BRANCH=
LIBHTP_REPO=
LIBHTP_BRANCH=

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 23011

File: doc/userguide/lua/lua-functions.rst [hunk 1]
```diff
@@ -6,7 +6,8 @@ Lua functions
 Differences between `output` and `detect`:
 ------------------------------------------
 
-Currently, the ``needs`` key initialization varies, depending on what is the goal of the script: output or detection.
+Currently, the ``needs`` key initialization varies, depending on what is the goal of the script: output or detection. The
+Lua script for the ``luaxform`` transform **does not use ``needs``**.
```
Comment (victorjulien) [inline doc/userguide/lua/lua-functions.rst:10]: PR example does show needs

File: doc/userguide/rules/transforms.rst [hunk 2]
```diff
@@ -243,3 +243,110 @@ This example transforms `"Zm 9v Ym Fy"` to `"foobar"`::
...
+    file.data; luaxform:./lua/lua-transform.lua, bytes 12, offset 13; content: "abc"; sid: 1;)
+
+The following Lua script shows a transform example that handles arguments: `bytes` and `offset` and uses
+those values (or defaults, if there are no arguments) for applying the uppercase transform to
+the buffer.
+
+.. code-block:: lua
+
+    function init (args)
+        local needs = {}
```
Comment (victorjulien) [inline doc/userguide/rules/transforms.rst:303]: not used according to doc earlier in PR?

File: src/detect-transform-luaxform.c [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+        goto early_return;
+    }
+
+    lua_newtable(luastate); /* stack at -1 */
+    if (lua_gettop(luastate) == 0 || lua_type(luastate, 2) != LUA_TTABLE) {
+        SCLogError("no table setup");
+        goto error;
+    }
+
+    lua_pushliteral(luastate, "script_api_ver"); /* stack at -2 */
```
Comment (victorjulien) [inline src/detect-transform-luaxform.c:118]: think we need to define what to do with this, if anything?

File: src/detect-transform-luaxform.c [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+error:
+    if (lua != NULL)
+        DetectTransformLuaxformFree(de_ctx, lua);
+    return NULL;
+}
+
+static void *DetectLuaxformThreadInit(void *data)
+{
+    DetectLuaxformData *lua = (DetectLuaxformData *)data;
+    BUG_ON(lua == NULL);
```
Comment (victorjulien) [inline src/detect-transform-luaxform.c:193]: handle error

File: src/detect.h [hunk 5]
```diff
@@ -371,6 +371,7 @@ struct DetectEngineThreadCtx_;// DetectEngineThreadCtx;
  * Prefilter and inspection will only deal with 'inspect'. */
 
 typedef struct InspectionBuffer {
+    struct DetectEngineThreadCtx_ *det_ctx;
```
Comment (victorjulien) [inline src/detect.h:374]: can we avoid this and pass it around as an argument?

File: src/util-magic.c [hunk 6]
```diff
@@ -120,11 +120,13 @@ char *MagicThreadLookup(magic_t *ctx, const uint8_t *buf, uint32_t buflen)
 
 #ifdef UNITTESTS
 
+#if 0
```
Comment (victorjulien) [inline src/util-magic.c:123]: why is this needed?

File: src/util-spm.c [hunk 7]
```diff
@@ -403,6 +400,9 @@ static uint8_t *BoyerMooreNocaseWrapper(uint8_t *text, uint8_t *in_needle, int t
 }
 
 #ifdef ENABLE_SEARCH_STATS
+/* Number of times to repeat the search (for stats) */
```
Comment (victorjulien) [inline src/util-spm.c:403]: is this related to the lua work?

**Review by victorjulien — CHANGES_REQUESTED** [review]
see inline

File: src/util-spm.c (hunk 7 above)
Comment (jlucovsky) [reply src/util-spm.c:403]: No;  There were warnings issued while rebuilding the Lua changes so I included that commit with the Lua changes.

File: doc/userguide/lua/lua-functions.rst (hunk 1 above)
Comment (jlucovsky) [reply doc/userguide/lua/lua-functions.rst:10]: Question: I'll make changes to eliminate the mention of `needs` in the test/doc content.

Should we retain the `init` function and call it during initialization and *not use/check* for a return falue?

File: src/util-spm.c (hunk 7 above)
Comment (jlucovsky) [reply src/util-spm.c:403]: I've submitted the fixup changes in another PR and will drop them.

**Comment by jlucovsky** [comment]
Continued in #11930
